### ISSUE TITLE
Redirect 504 timeout

### ISSUE THREAD

**Issue by javierin** [YOURLS/YOURLS#2074]
I have recently switched my web host for this install so i'm pretty sure this is environment related. Nginx config has been copied exactly from previous host to new one (updating service IP addresses only) However, I'm not sure of where to look at, because all i'm getting is 504 timeout errors.

I have a previously working YOURLS install in which I have configured to have a fallback URL.

If I go to a non existing redirect entry, let's say
https://redir.es/google12345678

It's redirecting me properly to the fallback URL, so I assume database access is working. I can also load the admin interface and create new redirection entries flawlessly.

However, If I try to visit an existing, correctly created (and present in the database) redirect link, I get a 504 timeout error.

https://redir.es/google

```
+---------+----------------------+--------+---------------------+--------------+--------+
| keyword | url                  | title  | timestamp           | ip           | clicks |
+---------+----------------------+--------+---------------------+--------------+--------+
| google  | http://www.google.es | Google | 2016-03-28 07:04:38 | <ip.address.edited> |      4 |
+---------+----------------------+--------+---------------------+--------------+--------+
```

I can also see that click counter for the urls is being increased, despite the 504 error.

I've been able to reproduce it both with chrome browser and curl, so i'm pretty sure it is not browser related.

### Technical details

* YOURLS version: 1.7.1 (also tried checking out master)
* PHP version: 5.6.11
* php5-mysqlnd 5.6.11
* nginx 1.9.3

**Comment by ozh** [comment]
From what you're telling (and that's a properly detailed issue, thanks) it seems that your problem happens here: https://github.com/YOURLS/YOURLS/blob/master/yourls-go.php#L26-L29, either on the line that updates detailed stats, or on the one that redirects

In your YOURLS root dir (next to `yourls-go.php` for instance) make a `test.php` file with the following:
```php
<?php
// skip various checks
define( 'YOURLS_FLOOD_DELAY_SECONDS', 0 );
define( 'YOURLS_INSTALLING', true );
define( 'YOURLS_NO_VERSION_CHECK', true );
define('YOURLS_PRIVATE', false); // No login stuff
require_once( dirname(__FILE__).'/includes/load-yourls.php' );
require_once( dirname(__FILE__).'/includes/functions-install.php' );
yourls_add_filter( 'html_title', create_function('','return "TEST";'));
yourls_html_head();
?>
<pre>
<?php

// test stuff here, var_dump things
// for instance here, start with testing the redirection bit:
yourls_redirect( 'http://ozh.org/', 301 );
```

Then load it in your browser to check how it works. If that works, test the other line. If that doesn't work, dig deeper into `your_redirect()` function to pin point the precise culprit, and so on.

To help you find where a function is located and dig deeper into its code, you can use http://yourls.org/xref/_functions/index.html

Feel free to comment on with further discoveries ;)

**Comment by javierin** [comment]
Ok, thanks to your detailed help i've been able to troubleshoot a litte bit more.

Your suggested test.php works, so I started to dig deeper into functions.php -> yourls_log_redirect.

Commenting the following line makes the redirection work:

$pre = yourls_apply_filter( 'shunt_log_redirect', false, $keyword );

It says it allows plugins to short circuit the whole function, so I started to look at the plugins that I am running:

Allow hypens
Don't encode/decode
Fallback URL
Piwik

I've tried disabling one at a time and finally got Piwik plugin as the switching factor between working / non working redirect.

Piwik plugin was trying to call piwik to record a redirect event. As the piwik instance is hosted in this same server and behind a nat, it was having issues reaching the public IP. Adding a static entry in /etc/hosts for the piwik domain -> private ip has solved this thing.

Again, thanks for your help ;)
